### PR TITLE
UnifiedStorage: Add ResourceTable format 

### PR DESCRIPTION
**Pull request by ryantxu**
While looking into the the search interface, one of the nagging issues is how do we return results -- PartialObjectMetadata is OK for just the reference, but to get real values there is no clear choice.

The plugins-sdk data.Frame is one option -- it has the properties we want with flexible columns and a known format.  BUT it is a pretty complicated format and does not have 1st class support in the apiserver (yet!).  The advantages of a columnar format are not useful for our needs here; and are honestly pretty difficult to work with.

BUT, we could build something that would resolve to [metav1.Table ](https://github.com/kubernetes/kubernetes/blob/v1.31.2/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1387)in k8s.  This has the HUGE advantage that most clients already understand it 🎉 

This could also be a more strict/field oriented target for the DocumentBuilder( https://github.com/grafana/grafana/pull/96281) and the current [SpecFieldMapping](https://github.com/grafana/grafana/blob/3ab608640ba55dc689ed514b7bcd5d48ed6898e3/pkg/storage/unified/resource/index_mapping.go#L158)

### CODE REVIEW

File: pkg/storage/unified/resource/resource.proto [hunk 1]
```diff
@@ -108,24 +108,24 @@ message ErrorDetails {
...
-	// Arrays are zero-indexed.  Fields may appear more than once in an array of
-	// causes due to fields having multiple errors.
-	// Optional.
-	//
-	// Examples:
-	//   "name" - the field "name" on the current resource
-	//   "items[0].name" - the field "name" on the first array entry in "items"
-	// +optional
-	string field = 3;
+  // A machine-readable description of the cause of the error. If this value is
```
Comment (ryantxu) [inline pkg/storage/unified/resource/resource.proto:111]: tabs to spaces

File: pkg/storage/unified/resource/table.go [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+
+	// Used for array indexing
+	reader func(iter *jsoniter.Iterator) (any, error)
+	writer func(v any, stream *jsoniter.Stream) error
+
+	OpenAPIType   string
+	OpenAPIFormat string
+}
+
+// nolint:gocyclo
```
Comment (ryantxu) [inline pkg/storage/unified/resource/table.go:171]: this is the complicated function -- for each type it picks how to encode/decode efficiently

There is likely a cleaner way to structure this, but it is all internal so maybe good enough to work with for now?

**Review by scottlepp — APPROVED** [review]
Curious how we should return faceted results.

File: pkg/storage/unified/resource/resource.proto [hunk 3]
```diff
@@ -448,6 +448,118 @@ message HealthCheckResponse {
...
+message ResourceTable {
+  // Columns describes each column in the returned items array. The number of cells per row
+  // will always match the number of column definitions.
+  repeated ResourceTableColumnDefinition columns = 1;
+
+  // rows is the list of items in the table.
+  repeated ResourceTableRow rows = 2;
+
+  // When more results exist, pass this in the next request
+  string next_page_token = 3;
```
Comment (owensmallwood) [inline pkg/storage/unified/resource/resource.proto:462]: do we need these last 3 fields? Looks like they're already part of `ListMeta` in the table.

Comment (owensmallwood) [reply pkg/storage/unified/resource/resource.proto:462]: nvm!

File: pkg/storage/unified/resource/table.go [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+type TableBuilder struct {
+	ResourceTable
+
+	lookup map[string]*resourceTableColumn
+
+	// Just keep track of it
+	hasDuplicateNames bool
+
+	// When add row gets a colum
+	ignoreUnknownColumns bool
```
Comment (owensmallwood) [inline pkg/storage/unified/resource/table.go:105]: whats the use case for turning this on?

**Review by owensmallwood — APPROVED** [review]
LGTM 🚀

File: pkg/storage/unified/resource/table.go (hunk 4 above)
Comment (ryantxu) [reply pkg/storage/unified/resource/table.go:105]: not sure... removed it :)

**Review by chaudyg — APPROVED** [review]
This is going to work great

**Pull request merged**